### PR TITLE
Update failing policy in Central Management fetcher and license checker if hit ES down node

### LINKED ISSUE #12776: Logstash crash when `xpack.management.elasticsearch.hosts` contains an unreachable host
When Logstash is configured to use Central Management and `xpack.management.elasticsearch.hosts` list contains an host that is down or unreachable, the process exits. The problem is that if the Elasticsearch management cluster has one node down, Logstash can't start at least the user manually remove that host from the above list. This is not a good user experience and requires manual intervention on `config/logstash.yml` that is not always feasible.

The idea could be to use a retry a couple of times and in case avoid to use the broken nodes in the list, let the Logstash process to start safely.
 
- Version:
- Operating System:
- Config File:
  - docker compose file 
[es_and_kb.yml.txt](https://github.com/elastic/logstash/files/6203003/es_and_kb.yml.txt)
   - section to add in `config/logstash.yml`
```
xpack.management.enabled: true
xpack.management.pipeline.id: ["main", "test*"]
xpack.management.elasticsearch.username: elastic
xpack.management.elasticsearch.password: changeme
xpack.management.elasticsearch.hosts: ["http://localhost:9200", "http://localhost:19200"]
```
- Steps to Reproduce:
   - run the docker file to bring up an Elasticsearch and Kibana configured to manage Logstash (`docker-copmpose -f  es_and_kb.yml up`
   - run Logstash `bin/logstash`
   - check the exception on the output

*Logstash output*
```
[2021-03-25T08:56:49,744][INFO ][logstash.runner          ] Log4j configuration path used is: /home/andrea/workspace/logstash_andsel/config/log4j2.properties
[2021-03-25T08:56:49,751][INFO ][logstash.runner          ] Starting Logstash {"logstash.version"=>"8.0.0", "jruby.version"=>"jruby 9.2.16.0 (2.5.7) 2021-03-03 f82228dc32 OpenJDK 64-Bit Server VM 11.0.7+10 on 11.0.7+10 +indy +jit [linux-x86_64]"}
[2021-03-25T08:56:49,801][INFO ][logstash.configmanagement.bootstrapcheck] Using Elasticsearch as config store {:pipeline_id=>["main", "test*"], :poll_interval=>"TimeValue{duration=5, timeUnit=SECONDS}ns"}
[2021-03-25T08:56:50,564][INFO ][logstash.configmanagement.elasticsearchsource] Configuration Management License OK
[2021-03-25T08:56:51,221][INFO ][logstash.configmanagement.elasticsearchsource] Elasticsearch pool URLs updated {:changes=>{:removed=>[], :added=>[http://elastic:xxxxxx@localhost:9200/, http://elastic:xxxxxx@localhost:19200/]}}
[2021-03-25T08:56:51,228][WARN ][logstash.configmanagement.elasticsearchsource] Restored connection to ES instance {:url=>"http://elastic:xxxxxx@localhost:9200/"}
[2021-03-25T08:56:51,232][INFO ][logstash.configmanagement.elasticsearchsource] ES Output version determined {:es_version=>7}
[2021-03-25T08:56:51,232][WARN ][logstash.configmanagement.elasticsearchsource] Detected a 6.x and above cluster: the `type` event field won't be used to determine the document _type {:es_version=>7}
[2021-03-25T08:56:51,237][WARN ][logstash.configmanagement.elasticsearchsource] Attempted to resurrect connection to dead ES instance, but got an error. {
	:url=>"http://elastic:xxxxxx@localhost:19200/", 
	:error_type=>LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError, 
	:error=>"Elasticsearch Unreachable: [http://elastic:xxxxxx@localhost:19200/][Manticore::SocketException] Connection refused (Connection refused)"}
[2021-03-25T08:56:51,249][WARN ][logstash.configmanagement.elasticsearchsource] Marking url as dead. Last error: [LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError] Elasticsearch Unreachable: [http://elastic:xxxxxx@localhost:19200/][Manticore::SocketException] Connection refused (Connection refused) {:url=>http://elastic:xxxxxx@localhost:19200/, :error_message=>"Elasticsearch Unreachable: [http://elastic:xxxxxx@localhost:19200/][Manticore::SocketException] Connection refused (Connection refused)", :error_class=>"LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError"}
[2021-03-25T08:56:51,255][ERROR][logstash.config.sourceloader] Could not fetch all the sources {:exception=>LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError, 
	:message=>"Elasticsearch Unreachable: [http://elastic:xxxxxx@localhost:19200/][Manticore::SocketException] Connection refused (Connection refused)", 
	:backtrace=>[
		"/home/andrea/workspace/logstash_andsel/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-10.8.4-java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:319:in `perform_request_to_url'",
		"/home/andrea/workspace/logstash_andsel/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-10.8.4-java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:304:in `block in perform_request'",
		"/home/andrea/workspace/logstash_andsel/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-10.8.4-java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:399:in `with_connection'",
		"/home/andrea/workspace/logstash_andsel/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-10.8.4-java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:303:in `perform_request'",
		"/home/andrea/workspace/logstash_andsel/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-10.8.4-java/lib/logstash/outputs/elasticsearch/http_client/pool.rb:311:in `block in Pool'",
		"/home/andrea/workspace/logstash_andsel/vendor/bundle/jruby/2.5.0/gems/logstash-output-elasticsearch-10.8.4-java/lib/logstash/outputs/elasticsearch/http_client.rb:162:in `get'",
		"/home/andrea/workspace/logstash_andsel/x-pack/lib/config_management/elasticsearch_source.rb:54:in `get_pipeline_fetcher'",
		"/home/andrea/workspace/logstash_andsel/x-pack/lib/config_management/elasticsearch_source.rb:80:in `pipeline_configs'",
		"/home/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/config/source_loader.rb:76:in `block in fetch'",
		"org/jruby/RubyArray.java:2572:in `collect'",
		"/home/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/config/source_loader.rb:75:in `fetch'",
		"/home/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/agent.rb:181:in `converge_state_and_update'",
		"/home/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/agent.rb:119:in `execute'",
		"/home/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/runner.rb:397:in `block in execute'",
		"/home/andrea/workspace/logstash_andsel/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/task.rb:24:in `block in initialize'"]}
```

### PR DESCRIPTION
**Pull request by andsel**
## Release notes

Fix bug in retrieving pipeline configuration from central management and license validation which make Logstash process crash if a configured node of the Elasticsearch cluster is down.

## What does this PR do?
Wraps the calls to the central management Elasticsearch cluster with the utility class `Stud::Try` to handle the remote  host error when the client used to connect hit a not available node.

## Why is it important/What is the impact to the user?

Central Management Elasticsearch node list is configured in `logstash.yml` file. If any of those Elasticsearch nodes is down when Logstash starts and hits one of down nodes then it could be that the license check fails and crash Logstash or that the not handled host error exception avoid to receive the pipelines updates.
With this fix the maintenance of an configuration cluster node doesn't prevent the Logstash start.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [x] Test the PR as described in the issue it resolves.

## How to test this PR locally

Follow the instructions provided in the resolved issue #12776.

## Related issues

- Fixes #12776

### CODE REVIEW

File: x-pack/lib/config_management/elasticsearch_source.rb [hunk 1]
```diff
@@ -53,7 +53,17 @@ def config_conflict?
...
+        loop do
+          begin
+            response = client.get("/")
+            break
+          rescue ::LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError => e
+            # continue looping until the client connects to a good host or NoConnectionAvailableError is raised.
+            failed_host = e.url
+            logger.info("Reiterate with next host, client should remove from the pool ", {:host => failed_host})
+          end
+        end
```
Comment (yaauie) [inline x-pack/lib/config_management/elasticsearch_source.rb:66]: We could use `Stud::try`, which is already available and supplies us with exponential backoff.

```suggestion
        response = Stud::try(loop, ::LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError) { client.get("/") }
```

If we want a better handler, we can subclass `Stud::Try`:

~~~ ruby
class ElasticsearchRequestHandler < Stud::Try
  def initialize(logger, name)
    @logger = logger
    @name = name
  end

  def log_failure(exception, fail_count, message)
    @logger.warn("Attempt to #{@name} failed. #{message}", fail_count: fail_count, exception: e.message)
  end
end
~~~

Which would make the above:

```suggestion
        retry_handler = ElasticsearchRequestHandler.new(logger, 'Elasticsearch capability detection')
        response = retry_handler.try(loop, ::LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError) { client.get("/") }
```

Comment (andsel) [reply x-pack/lib/config_management/elasticsearch_source.rb:66]: Thanks @yaauie, nice suggestion

**Comment by andsel** [comment]
Jenkins test this please

**Comment by andsel** [comment]
Jenkins test this

**Comment by andsel** [comment]
Jenkins test this

**Comment by andsel** [comment]
Ping @yaauie here because the review is in pending state.

File: x-pack/spec/license_checker/license_reader_spec.rb [hunk 2]
```diff
@@ -51,22 +51,40 @@
...
+        first_call = true
+        allow(mock_client).to receive(:get) do |path|
+          expect(path).to eq('_xpack')
+          if first_call
+            first_call = false
+            raise host_not_reachable
+          end
+          xpack_info
+        end
+      end
```
Comment (yaauie) [inline x-pack/spec/license_checker/license_reader_spec.rb:83]: We can rely on multiple ordered expectations on our partial mock, which should make it more readable.

```suggestion
      before(:each) do
        # set up expectation of single failure
        expect(license_reader.logger).to receive(:warn).with(a_string_including "Attempt to validate Elasticsearch license failed.")
        expect(mock_client).to receive(:get).with('_xpack').and_raise(host_not_reachable).once
        
        # ensure subsequent success
        expect(mock_client).to receive(:get).with('_xpack').and_return(xpack_info)
      end
```

**Review by yaauie — COMMENTED** [review]
Implementation looks good, and should provide some safety 👍🏼 

One nitpick about using rspec's partial mocks -- expectations that fail inside a mock implementation can be particularly difficult to read and debug.

File: x-pack/spec/license_checker/license_reader_spec.rb (hunk 2 above)
Comment (andsel) [reply x-pack/spec/license_checker/license_reader_spec.rb:83]: You are right, far better, I missed I could instruct multiple responses for same call.

**Comment by andsel** [comment]
Jenkins test this please

**Comment by andsel** [comment]
Jenkins test this please

**Comment by yaauie** [comment]
jenkins test this again please

**Review by yaauie — COMMENTED** [review]
Changes LGTM, but I would prefer a green build before merging.

**Review by yaauie — APPROVED** [review]

**Comment by andsel** [comment]
@logstashmachine backport 7.17

**Pull request merged**